custom_components/localtuya/light.py

```diff
@@ -141,31 +167,66 @@ async def async_turn_on(self, **kwargs):
                 self._upper_brightness,
             )
             if self._is_white_mode:
-                await self._device.set_dp(brightness, self._config.get(CONF_BRIGHTNESS))
+                states[self._config.get(CONF_BRIGHTNESS)] = brightness
+
             else:
-                color = "{:04x}{:04x}{:04x}".format(
-                    round(self._hs[0]), round(self._hs[1] * 10.0), brightness
-                )
-                await self._device.set_dp(color, self._config.get(CONF_COLOR))
+                if self._is_rgb_mode:
+                    rgb = color_util.color_hsv_to_RGB(
+                        self._hs[0],
+                        self._hs[1],
+                        int(brightness * 100 / self._upper_brightness),
+                    )
+                    color = "{:02x}{:02x}{:02x}{:04x}{:02x}{:02x}".format(
+                        round(rgb[0]),
+                        round(rgb[1]),
+                        round(rgb[2]),
+                        round(self._hs[0]),
+                        round(self._hs[1] * 255 / 100),
+                        brightness,
+                    )
+                else:
+                    color = "{:04x}{:04x}{:04x}".format(
+                        round(self._hs[0]), round(self._hs[1] * 10.0), brightness
+                    )
+                states[self._config.get(CONF_COLOR)] = color
 
         if ATTR_HS_COLOR in kwargs and (features & SUPPORT_COLOR):
             hs = kwargs[ATTR_HS_COLOR]
-            color = "{:04x}{:04x}{:04x}".format(
-                round(hs[0]), round(hs[1] * 10.0), self._brightness
-            )
-            await self._device.set_dp(color, self._config.get(CONF_COLOR))
+            if self._is_rgb_mode is None:
+                self._is_rgb_mode = len(self.dps_conf(CONF_COLOR)) > 12
```

Comment:
Am I understanding it right that `_is_rgb_mode` really represents if the bulb supports RGB or not and not current mode of the lamp (which I'm doing something similar for with white color)? It's a bit confusing to have two properties named the opposite of each other and not really representing the same thing. Can we consolidate into one variable?

Reply:
And it also seems strange to assign this variable here, would make more sense to set in `status_updated` in case the value change. I would however never expect it to do so, so it could possibly be assigned directly in the constructor.

Reply:
This one is representing the color coding, RGB_HSV if True, HSV4Bytes if False

Reply:
I set it originally in  `status_updated`. It seems that it's not called when HA is restarted until something changed. If we have a bulb already On in white and just change color, the coding is not yet defined.

I may try to set it in the constructor but need to ensure data is already there. The problem with bulbs is that they can be powered off and not accessible when the constructor is called. In that case, we do not have the necessary data to update the value.

Reply:
Eventually, i extracted the check in a method without using variable.